### PR TITLE
feat: daily streak calculation

### LINKED ISSUE #70: Daily Streak Calculation Service
## Task Description
Implement service to calculate current consecutive daily completion streaks for habits.

## Acceptance Criteria
- [ ] StreakService calculates current streaks for daily habits
- [ ] Handles timezone-aware date calculations correctly
- [ ] Comprehensive tests cover various scenarios and edge cases
- [ ] Caching implemented for performance optimization

## Technical Notes
- Calculate consecutive days from most recent check-in backwards
- Handle gaps in check-ins and timezone changes properly
- Document streak calculation algorithm (consecutive days without gaps)

### PR DESCRIPTION
**Pull request by hyuseinleshov**
## Summary
Implements a daily streak calculation feature with Redis caching for habit check-ins. The system calculates consecutive days of check-ins from the most recent backwards, using timezone-aware date calculations to ensure accuracy across different user timezones.

## Issue
Closes #70

## Testing
- [ ] Unit tests added/updated
- [ ] Integration tests added/updated
- [x] Manual testing completed
- [x] All tests pass

## Checklist
- [x] Code builds successfully
- [x] Documentation updated (if needed)
- [x] No breaking changes (or documented)
- [x] Code review requested

## Notes for Reviewers
**Caching Strategy:**
- Cache hit = O(1) Redis lookup (no DB query)
- Cache miss = full calculation + store in Redis
- Cache TTL: expires at midnight of the day after tomorrow (user's timezone)
- Cache invalidation: on check-in creation (increments cached value) and deletion (recalculates)

**Design Decisions:**
1. **Timezone-aware calculations**: Uses user's timezone for accurate date boundaries
2. **Descending order optimization**: Database index on `(habit_id, created_at DESC)` improves query performance for long streaks
3. **Flyway integration**: Database migrations now managed via Flyway for version control

### CODE REVIEW

File: src/main/java/com/habittracker/api/checkin/service/impl/CheckInServiceImpl.java [hunk 1]
```diff
@@ -48,6 +53,13 @@ public CheckInResponse checkIn(UUID habitId, UserEntity user) {
     CheckInEntity checkInEntity = new CheckInEntity();
     checkInEntity.setHabit(habit);
     CheckInEntity saved = checkInRepository.save(checkInEntity);
+
+    String cacheKey = STREAK_CACHE_KEY_PREFIX + habitId;
+    Integer currentStreak = (Integer) redisTemplate.opsForValue().get(cacheKey);
+    if (currentStreak != null) {
+      redisTemplate.opsForValue().increment(cacheKey);
+    }
```
Comment (ShenolShengov) [inline src/main/java/com/habittracker/api/checkin/service/impl/CheckInServiceImpl.java:61]: Finish the logic, what happen if not have streak for this habit in redis.

File: src/main/java/com/habittracker/api/checkin/service/impl/StreakServiceImpl.java [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+    log.debug("Streak cache miss for habit ID: {}", habitId);
+    log.debug("Calculating streak for habit ID: {}", habitId);
+
+    HabitEntity habit = habitHelper.getNotDeletedOrThrow(habitId);
+    ZoneId userTimeZone = parseTimeZone(habit.getUser().getUserProfile().getTimezone());
+
+    List<CheckInEntity> checkIns = checkInRepository.findByHabitIdOrderByCreatedAtDesc(habitId);
+
+    int currentStreak;
+    if (checkIns.isEmpty()) {
```
Comment (ShenolShengov) [inline src/main/java/com/habittracker/api/checkin/service/impl/StreakServiceImpl.java:56]: You have this check in method `calculateConsecutiveDays`

File: src/main/java/com/habittracker/api/checkin/service/impl/StreakServiceImpl.java [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+
+    // 3. Store in Redis with TTL until midnight of the day after tomorrow (user's timezone)
+    Duration untilMidnight = calculateDurationUntilMidnight(userTimeZone);
+    Duration cacheTtl = untilMidnight.plusDays(1);
+    redisTemplate.opsForValue().set(cacheKey, currentStreak, cacheTtl);
+
+    return new StreakResponse(habitId, currentStreak, Instant.now());
+  }
+
+  private int calculateConsecutiveDays(List<CheckInEntity> checkIns, ZoneId userTimeZone) {
```
Comment (ShenolShengov) [inline src/main/java/com/habittracker/api/checkin/service/impl/StreakServiceImpl.java:70]: You have to extract what date is today for user timezone and start checking for chek-ins in previous date from this date (if have check today or yesterday).

File: src/main/java/com/habittracker/api/checkin/service/StreakService.java [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+   * timezone 3. Starting from the most recent check-in, count consecutive days backwards 4. A
+   * streak is broken when there's a gap of more than one day between check-ins
+   *
+   * <p>Example: - Check-ins on: Jan 5, Jan 4, Jan 3, Jan 1 → Streak = 3 (Jan 5, 4, 3) - Check-ins
+   * on: Jan 5, Jan 3, Jan 2, Jan 1 → Streak = 1 (Jan 5 only) - No check-ins → Streak = 0
+   *
+   * @param habitId the UUID of the habit to calculate the streak for
+   * @return StreakResponse containing the current streak count and calculation timestamp
+   */
+  StreakResponse calculateStreak(UUID habitId);
```
Comment (ShenolShengov) [inline src/main/java/com/habittracker/api/checkin/service/StreakService.java:30]: Split this method to `incermentSterak` and `calculateSteak`. `incrementSteark` should depends on `calculateStreak`.

File: src/main/java/com/habittracker/api/checkin/service/impl/StreakServiceImpl.java [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+
+    // 2. Cache miss - calculate from database
+    log.debug("Streak cache miss for habit ID: {}", habitId);
+    log.debug("Calculating streak for habit ID: {}", habitId);
+
+    HabitEntity habit = habitHelper.getNotDeletedOrThrow(habitId);
+    ZoneId userTimeZone = parseTimeZone(habit.getUser().getUserProfile().getTimezone());
+
+    // 3. Check if the most recent check-in is from today or yesterday
+    CheckInEntity mostRecentCheckIn =
```
Comment (ShenolShengov) [inline src/main/java/com/habittracker/api/checkin/service/impl/StreakServiceImpl.java:54]: Use `Optional` efficient

**Review by ShenolShengov — CHANGES_REQUESTED** [review]
